lua/ttt2/extensions/net.lua

```diff
@@ -6,58 +6,201 @@
+-- @realm shared
+-- @internal
+local function sendNextStream(messageId, streamId, split)
+	net.Start(NETMSG_STREAM)
+	-- Write the messageId
+	net.WriteString(messageId)
+	-- Write the streamId
+	net.WriteUInt(streamId, 32)
+
+	local sendMoreSplits = split > 1
+	-- Write if there are still fragments coming after this one
+	net.WriteBool(sendMoreSplits)
+
+	local data = wait_stream_cache[messageId][streamId]
+	-- Write the actual data fragment as a string, which internally will also send its size
+	net.WriteString(data.splits[#data.splits + 1 - split])
+
+	if SERVER then
+		if data.plys then
+			net.Send(data.plys)
+		else
+			net.Broadcast()
+		end
+	else
+		net.SendToServer()
+	end
+
+	if not sendMoreSplits then
+		local confirmations = current_split_id[messageId]
+		confirmations[streamId] = 1
+
+		-- Check if all outstanding streams are handled before reset
+		for i = 1, #confirmations do
+			if confirmations[i] > 1 then return end
+		end
+
+		-- Reset all caches
+		wait_stream_cache[messageId] = {}
+		current_split_id[messageId] = {}
+		waiting_Plys[messageId] = {}
+		backup_Plys[messageId] = {}
+	end
+end
+
+---
+-- Sends the next split and removes overrideTimers if given
+-- @param string messageId a unique message id similar to the network strings
+-- @param number streamId the current stream number the split is requested for
+-- @param bool timerOverride if the timer called this function instead of the normal way
+local function SendNextSplit(messageId, streamId, timerOverride)
+	if timerOverride then
+		waiting_Plys[messageId][streamId] = table.Copy(backup_Plys[messageId][streamId])
+	else
+		local timerName = messageId .. streamId
+		if timer.Exists(timerName) then
+			timer.Remove(timerName)
+		end
+	end
+
+	local splitId = current_split_id[messageId][streamId] - 1
+	current_split_id[messageId][streamId] = splitId
+
+	sendNextStream(messageId, streamId, splitId)
+end
+
+---
+-- Checks if all players in the cached players list have received it before 
+-- it sends the next part of the stream
+-- @param number len Length of message in Bits
+-- @param Player ply player that the message has received, `nil` on the client
+-- @realm shared
+-- @internal
+local function CheckConfirmedSplit(len, ply)
+	local messageId = net.ReadString()
+	local streamId = net.ReadUInt(32)
+
+	-- Check if all plys of the given list have the info received
+	local cPlys = waiting_Plys[messageId] and waiting_Plys[messageId][streamId]
+	if cPlys then
+		local plyID64 = ply:SteamID64()
+		table.RemoveByValue(cPlys, plyID64)
+
+		if #cPlys > 0 then
+			local timerName = messageId .. streamId
+
+			-- If some players timeout or the given playerList is faulty, just send the next update after some time anyways
+			if not timer.Exists(timerName) then
+				timer.Create(timerName, 5, 1, function() SendNextSplit(messageId, splitId, true) end)
+			end
+
+			return
+		end
+
+		waiting_Plys[messageId][streamId] = table.Copy(backup_Plys[messageId][streamId])
+	end
```

Comment:
I would implement it per player and not create an artificial sync barrier and wait for clients.

Reply:
Sounds better, not sure right now why I did it like this in the first place.

Reply:
I did that and moved it to another PR.